reference-service/src/main/java/com/transformuk/hee/tis/reference/service/api/RoleResource.java

```diff
@@ -244,6 +244,46 @@ public RoleResource(RoleRepository roleRepository, RoleMapper roleMapper,
     return new ResponseEntity<>(result, HttpStatus.OK);
   }
 
+  /**
+   * EXISTS /roles/matches/ : check if there's a role in the database that matches the code
+   * provided, regardless of casing.
+   *
+   * @param codes            the codes of the RoleDTOs to check
+   * @param columnFilterJson The column filters to apply
+   * @return Map             Where a key is the code to be matched, and a value is the code that was
+   *     matched from the database.
+   */
+  @PostMapping("/roles/matches/")
+  public ResponseEntity<Map<String, String>> rolesMatch(@RequestBody List<String> codes,
+      @RequestParam(value = "columnFilters", required = false) String columnFilterJson)
+      throws IOException {
+    codes = codes.stream()
+        .map(code -> getConverter(code).decodeUrl().toString())
+        .collect(Collectors.toList());
+    log.debug("REST request to check Roles match: {}", codes);
+    Specification<Role> specs = Specification.where(in("code", new ArrayList<>(codes)));
+
+    List<Class> filterEnumList = Lists.newArrayList(Status.class);
+    List<ColumnFilter> columnFilters =
+        ColumnFilterUtil.getColumnFilters(columnFilterJson, filterEnumList);
+
+    for (ColumnFilter columnFilter : columnFilters) {
+      specs = specs.and(in(columnFilter.getName(), columnFilter.getValues()));
+    }
+
+    List<Role> roles = roleRepository.findAll(specs);
+
+    Set<String> foundCodes = roles.stream().map(Role::getCode).collect(Collectors.toSet());
+
+    Map<String, String> result = codes.stream()
```

Comment:
Here's a simpler example:
`curl -X POST http://local.tis.com/reference/api/roles/matches/ -H 'cache-control: no-cache' -H "Content-Type: application/json" -d '["ARCP/Recruitment Assessor","ARCP/Recruitment Assessor"]'`

Reply:
Well spotted. This appears to have always been a vulnerability all along since the old endpoint also returns a Map, and by passing two duplicate roles into a Person row it currently fails when it tries to put two duplicate keys in the Map (the error on the spreadsheet simply reads 'Internal Server Error').

I don't know whether it's better to:

1. Prevent duplicates from TCS, right before it calls the endpoint, so that two duplicates are never passed in
or
2. Make the endpoint duplicate-safe, so that it returns a specific exception that TCS will then have to handle

The first approach leaves the endpoint a bit vulnerable and leaves it up to the developers to use it sensibly (i.e. put checks in place in TCS before it passes duplicates to it).
The second approach makes the endpoint safer, so that there are less chances for anybody to misuse it, even in possible future developments.

I feel like the second one is the best approach, even if it makes it a bit harder to then capture that failure in TCS to handle it correctly when returning the PersonDto.

Reply:
I wonder if it should simply return a map containing the distinct values that were passed in, and not throw an exception? To me, that would be reasonable. If the calling process can't handle the fact that the map it's getting back is not the same length as the list it passed in (if there were duplicates in the list), then that would seem to me to be a problem for the calling process to sort out itself.

Reply:
So basically, (2) above, except not throwing an exception.

Reply:
Sorry for the late response, I don't seem to be getting alerts for these updates.

Ah, apologies - I should have been clearer when I made my suggestion. I actually thought this commit was perfect: https://github.com/Health-Education-England/TIS-REFERENCE/pull/202/commits/76b5e6f916fe8236a9166e865aa3d0a8aad6b7ea!

To me its slightly odd to filter out case-insensitive duplicates in the codes submitted, but to use the first one (with its casing intact) as the 'master' that is returned.

However, this is maybe getting into hair-splitting territory for some quite far-out edge cases. I'm happy to proceed with either version. If you want to go back to the previous commit let me know, otherwise I'll approve after stand-up.